lib/auth/webauthncli/fido2_test.go

```diff
@@ -0,0 +1,1438 @@
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "challenge",
+		},
+		{
+			name:   "cc without RPID",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.RelyingParty.ID = ""
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "relying party ID",
+		},
+		{
+			name:   "cc unsupported parameters",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.Parameters = []protocol.CredentialParameter{
+					{Type: protocol.PublicKeyCredentialType, Algorithm: webauthncose.AlgEdDSA},
+				}
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "ES256",
+		},
+		{
+			name:     "nil pinPrompt",
+			origin:   origin,
+			createCC: func() *wanlib.CredentialCreation { return okCC },
+			wantErr:  "prompt",
+		},
+		{
+			name:   "rrk empty RP name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.RelyingParty.Name = ""
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "relying party name",
+		},
+		{
+			name:   "rrk empty user name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.Name = ""
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "user name",
+		},
+		{
+			name:   "rrk empty user display name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.DisplayName = ""
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "user display name",
+		},
+		{
+			name:   "rrk nil user ID",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.ID = nil
+				return &cp
+			},
+			prompt:  prompt,
+			wantErr: "user ID",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Millisecond)
+			defer cancel()
+
+			_, err := wancli.FIDO2Register(ctx, test.origin, test.createCC(), test.prompt)
+			require.Error(t, err, "FIDO2Register returned err = nil, want %q", test.wantErr)
+			assert.Contains(t, err.Error(), test.wantErr, "FIDO2Register returned err = %q, want %q", err, test.wantErr)
+		})
+	}
+}
+
+func resetFIDO2AfterTests(t *testing.T) {
+	pollInterval := wancli.FIDO2PollInterval
+	devLocations := wancli.FIDODeviceLocations
+	newDevice := wancli.FIDONewDevice
+	t.Cleanup(func() {
+		wancli.FIDO2PollInterval = pollInterval
+		wancli.FIDODeviceLocations = devLocations
+		wancli.FIDONewDevice = newDevice
+	})
+}
+
+type fakeFIDO2 struct {
+	locs    []*libfido2.DeviceLocation
+	devices map[string]*fakeFIDO2Device
+}
+
+func newFakeFIDO2(devs ...*fakeFIDO2Device) *fakeFIDO2 {
+	f := &fakeFIDO2{
+		devices: make(map[string]*fakeFIDO2Device),
+	}
+	for _, dev := range devs {
+		if _, ok := f.devices[dev.path]; ok {
+			panic(fmt.Sprintf("Duplicate device path registered: %q", dev.path))
```

Comment:
I'd rather keep the fakes "clean" of `testing.T`. It's a bit of purity on my part, but it makes it easier in case we want to extract them to other places. It also seemed unnecessary to change the signature for what is essentially a sanity check.

Let me know if you feel strongly about this.

Reply:
I only really suggested this and the comment below because it seemed like this was test only code. Feel free to disregard